jOOQ-meta/src/main/java/org/jooq/meta/tools/FlywayFileComparator.java

```diff
@@ -0,0 +1,57 @@
+/*
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *  http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ *
+ * Other licenses:
+ * -----------------------------------------------------------------------------
+ * Commercial licenses for this work are available. These replace the above
+ * ASL 2.0 and offer limited warranties, support, maintenance, and commercial
+ * database integrations.
+ *
+ * For more information, please visit: http://www.jooq.org/licenses
+ *
+ *
+ *
+ *
+ *
+ *
+ *
+ *
+ *
+ *
+ *
+ *
+ *
+ *
+ *
+ *
+ */
+package org.jooq.meta.tools;
+
+import java.io.File;
+import java.util.Comparator;
+
+public final class FlywayFileComparator implements Comparator<File> {
+
+    public static final FlywayFileComparator INSTANCE = new FlywayFileComparator();
+
+    @Override
+    public final int compare(File o1, File o2) {
+        String s1 = o1 == null ? null : o1.getName();
+        String s2 = o2 == null ? null : o2.getName();
+
+        FlywayVersion v1 = FlywayVersion.fromVersion(s1 == null ? null : s1.substring(1, s1.indexOf("__")));
```

Comment:
introduce SEPARATOR for "__"

Reply:
This substring call can lead to:

```
Caused by: java.lang.StringIndexOutOfBoundsException: begin 1, end -1, length 12
    at java.lang.String.checkBoundsBeginEnd (String.java:3319)
    at java.lang.String.substring (String.java:1874)
    at org.jooq.meta.tools.FlywayFileComparator.compare (FlywayFileComparator.java:52)
    at org.jooq.meta.tools.FlywayFileComparator.compare (FlywayFileComparator.java:43)
```

e.g. when comparing non-flyway files or directories

Reply:
This seems to work better:

```java
public final class FlywayFileComparator implements Comparator<File> {

    public static final FlywayFileComparator INSTANCE = new FlywayFileComparator();

    @Override
    public final int compare(File o1, File o2) {
        String s1 = o1 == null ? null : o1.getName();
        String s2 = o2 == null ? null : o2.getName();

        if (s1 != null && s2 != null) {
            int i1 = s1.indexOf("__");
            int i2 = s2.indexOf("__");

            if (i1 != -1 && i2 != -1) {
                FlywayVersion v1 = FlywayVersion.fromVersion(s1.substring(1, i1));
                FlywayVersion v2 = FlywayVersion.fromVersion(s2.substring(1, i2));

                return v1.compareTo(v2);
            }
        }

        // In case we're not comparing two Flyway files, fall back to default comparator logic
        return FileComparator.INSTANCE.compare(o1, o2);
    }
}
```

Reply:
its not better: (i1 >=1 && i2 >=1), otherwise a File starting with "__" will still trigger an exception

Reply:
You're right. Are empty version numbers possible? Or should we check for `(i1 > 1 && i2 > 1)`?

Reply:
Never mind. The `FlywayVersion.fromVersion()` method already handles this.

Reply:
:+1: